File: apps/react-native-app/src/theme/ThemeTransitionManager.ts
//
// ThemeTransitionManager.ts
// Aether React Native App
//
// Dynamic theme transitions with smooth animations and system integration.
// This manager provides polished, delightful theme switching experiences
// that respect user preferences and accessibility settings.
//

import { useState, useEffect, useRef, useCallback } from 'react';
import { Animated, Easing, AccessibilityInfo } from 'react-native';
import { Theme } from './ThemeProvider';

// MARK: - Transition Types

/**
 * Types of theme transitions available
 */
export enum ThemeTransitionType {
  FADE = 'fade',
  SLIDE = 'slide',
  SCALE = 'scale',
  MORPH = 'morph',
  CROSSFADE = 'crossfade',
  DISSOLVE = 'dissolve',
}

export const TransitionTypeLabels: Record<ThemeTransitionType, string> = {
  [ThemeTransitionType.FADE]: 'Fade',
  [ThemeTransitionType.SLIDE]: 'Slide',
  [ThemeTransitionType.SCALE]: 'Scale',
  [ThemeTransitionType.MORPH]: 'Morph',
  [ThemeTransitionType.CROSSFADE]: 'Crossfade',
  [ThemeTransitionType.DISSOLVE]: 'Dissolve',
};

export const TransitionTypeDescriptions: Record<ThemeTransitionType, string> = {
  [ThemeTransitionType.FADE]: 'Smooth fade between themes',
  [ThemeTransitionType.SLIDE]: 'Slide transition with direction',
  [ThemeTransitionType.SCALE]: 'Scale and fade combination',
  [ThemeTransitionType.MORPH]: 'Morphing color transitions',
  [ThemeTransitionType.CROSSFADE]: 'Crossfade with blur effect',
  [ThemeTransitionType.DISSOLVE]: 'Dissolve with particle effect',
};

// MARK: - Animation Curves

/**
 * Predefined animation curves for theme transitions
 */
export enum ThemeAnimationCurve {
  EASE_IN_OUT = 'easeInOut',
  EASE_IN = 'easeIn',
  EASE_OUT = 'easeOut',
  SPRING = 'spring',
  BOUNCY = 'bouncy',
  SMOOTH = 'smooth',
}

export const AnimationCurveLabels: Record<ThemeAnimationCurve, string> = {
  [ThemeAnimationCurve.EASE_IN_OUT]: 'Ease In Out',
  [ThemeAnimationCurve.EASE_IN]: 'Ease In',
  [ThemeAnimationCurve.EASE_OUT]: 'Ease Out',
  [ThemeAnimationCurve.SPRING]: 'Spring',
  [ThemeAnimationCurve.BOUNCY]: 'Bouncy',
  [ThemeAnimationCurve.SMOOTH]: 'Smooth',
};

export const getAnimationConfig = (curve: ThemeAnimationCurve) => {
  switch (curve) {
    case ThemeAnimationCurve.EASE_IN_OUT:
      return {
        duration: 600,
        easing: Easing.inOut(Easing.ease),
      };
    case ThemeAnimationCurve.EASE_IN:
      return {
        duration: 500,
        easing: Easing.in(Easing.ease),
      };
    case ThemeAnimationCurve.EASE_OUT:
      return {
        duration: 500,
        easing: Easing.out(Easing.ease),
      };
    case ThemeAnimationCurve.SPRING:
      return {
        duration: 600,
        easing: Easing.out(Easing.back(1.2)),
      };
    case ThemeAnimationCurve.BOUNCY:
      return {
        duration: 400,
        easing: Easing.out(Easing.back(1.5)),
      };
    case ThemeAnimationCurve.SMOOTH:
      return {
        duration: 800,
        easing: Easing.inOut(Easing.cubic),
      };
  }
};

// MARK: - Transition Direction

/**
 * Slide transition directions
 */
export enum TransitionDirection {
  LEFT = 'left',
  RIGHT = 'right',
  UP = 'up',
  DOWN = 'down',
}

export const DirectionLabels: Record<TransitionDirection, string> = {
  [TransitionDirection.LEFT]: 'Left',
  [TransitionDirection.RIGHT]: 'Right',
  [TransitionDirection.UP]: 'Up',
  [TransitionDirection.DOWN]: 'Down',
};

// MARK: - Transition State

/**
 * State tracking for theme transitions
 */
export interface ThemeTransitionState {
  isTransitioning: boolean;
  transitionProgress: Animated.Value;
  currentTransitionType: ThemeTransitionType;
  currentAnimationCurve: ThemeAnimationCurve;
  transitionDirection: TransitionDirection;
  shouldReduceMotion: boolean;
  shouldReduceTransparency: boolean;
}

// MARK: - Theme Transition Manager

/**
 * Manager for handling smooth theme transitions with animations
 */
export class ThemeTransitionManager {
  private transitionState: ThemeTransitionState;
  private previousTheme: Theme | null = null;
  private nextTheme: Theme | null = null;
  private animationRef = useRef<Animated.CompositeAnimation | null>(null);
  private accessibilityListeners: Array<() => void> = [];

  constructor() {
    this.transitionState = {
      isTransitioning: false,
      transitionProgress: new Animated.Value(0),
      currentTransitionType: ThemeTransitionType.FADE,
      currentAnimationCurve: ThemeAnimationCurve.EASE_IN_OUT,
      transitionDirection: TransitionDirection.RIGHT,
      shouldReduceMotion: false,
      shouldReduceTransparency: false,
    };

    this.setupAccessibilityListeners();
  }

  // MARK: - Setup

  private setupAccessibilityListeners() {
    // Listen for reduce motion changes
    const reduceMotionListener = AccessibilityInfo.addEventListener(
      'reduceMotionChanged',
      this.handleReduceMotionChange
    );

    // Listen for reduce transparency changes
    const reduceTransparencyListener = AccessibilityInfo.addEventListener(
      'reduceTransparencyChanged',
      this.handleReduceTransparencyChange
    );

    this.accessibilityListeners.push(() => reduceMotionListener?.remove());
    this.accessibilityListeners.push(() => reduceTransparencyListener?.remove());

    // Get initial accessibility settings
    this.updateAccessibilitySettings();
  }

  private handleReduceMotionChange = (isReduceMotionEnabled: boolean) => {
    this.transitionState.shouldReduceMotion = isReduceMotionEnabled;
    this.updateAccessibilitySettings();
  };

  private handleReduceTransparencyChange = (isReduceTransparencyEnabled: boolean) => {
    this.transitionState.shouldReduceTransparency = isReduceTransparencyEnabled;
    this.updateAccessibilitySettings();
  };

  private async updateAccessibilitySettings() {
    try {
      const isReduceMotionEnabled = await AccessibilityInfo.isReduceMotionEnabled();
      const isReduceTransparencyEnabled = await AccessibilityInfo.isReduceTransparencyEnabled();

      this.transitionState.shouldReduceMotion = isReduceMotionEnabled;
      this.transitionState.shouldReduceTransparency = isReduceTransparencyEnabled;

      // Adjust transition behavior based on accessibility settings
      if (isReduceMotionEnabled) {
        this.transitionState.currentAnimationCurve = ThemeAnimationCurve.EASE_IN_OUT;
      }
    } catch (error) {
      console.warn('Failed to get accessibility settings:', error);
    }
  }

  // MARK: - Public Methods

  /**
   * Perform a smooth theme transition
   */
  public transitionToTheme = (
    currentTheme: Theme,
    newTheme: Theme,
    type: ThemeTransitionType = ThemeTransitionType.FADE,
    curve: ThemeAnimationCurve = ThemeAnimationCurve.EASE_IN_OUT,
    direction: TransitionDirection = TransitionDirection.RIGHT,
    onComplete?: () => void
  ) => {
    if (this.transitionState.isTransitioning) {
      return;
    }

    // Store transition parameters
    this.previousTheme = currentTheme;
    this.nextTheme = newTheme;
    this.transitionState.currentTransitionType = type;
    this.transitionState.currentAnimationCurve = curve;
    this.transitionState.transitionDirection = direction;

    // Check accessibility settings
    if (this.transitionState.shouldReduceMotion) {
      // Instant transition for reduced motion
      this.performInstantTransition(newTheme, onComplete);
    } else {
      // Animated transition
      this.performAnimatedTransition(newTheme, type, curve, onComplete);
    }
  };

  /**
   * Get current transition state
   */
  public getTransitionState(): ThemeTransitionState {
    return this.transitionState;
  }

  /**
   * Get transition progress value for animations
   */
  public getTransitionProgress(): Animated.Value {
    return this.transitionState.transitionProgress;
  }

  /**
   * Check if currently transitioning
   */
  public isTransitioning(): boolean {
    return this.transitionState.isTransitioning;
  }

  /**
   * Get interpolated color between two themes
   */
  public interpolateColor(
    startColor: string,
    endColor: string,
    progress: number
  ): string {
    // Simple color interpolation - in a real implementation,
    // you'd want to interpolate RGB/HSB values properly
    return progress > 0.5 ? endColor : startColor;
  }

  /**
   * Cleanup resources
   */
  public cleanup() {
    this.accessibilityListeners.forEach(listener => listener());
    this.accessibilityListeners = [];
    
    if (this.animationRef.current) {
      this.animationRef.current.stop();
      this.animationRef.current = null;
    }
  }

  // MARK: - Private Methods

  private performInstantTransition(newTheme: Theme, onComplete?: () => void) {
    this.transitionState.isTransitioning = true;
    this.transitionState.transitionProgress.setValue(1);

    // Apply theme change immediately
    setTimeout(() => {
      this.transitionState.isTransitioning = false;
      this.transitionState.transitionProgress.setValue(0);
      this.previousTheme = null;
      this.nextTheme = null;
      onComplete?.();
    }, 100);
  }

  private performAnimatedTransition(
    newTheme: Theme,
    type: ThemeTransitionType,
    curve: ThemeAnimationCurve,
    onComplete?: () => void
  ) {
    this.transitionState.isTransitioning = true;
    this.transitionState.transitionProgress.setValue(0);

    const config = getAnimationConfig(curve);

    // Start transition animation
    this.animationRef.current = Animated.timing(
      this.transitionState.transitionProgress,
      {
        toValue: 1,
        duration: config.duration,
        easing: config.easing,
        useNativeDriver: true,
      }
    );

    this.animationRef.current.start((result) => {
      if (result.finished) {
        this.completeTransition(onComplete);
      }
    });
  }

  private completeTransition(onComplete?: () => void) {
    this.transitionState.isTransitioning = false;
    
    // Reset progress
    setTimeout(() => {
      this.transitionState.transitionProgress.setValue(0);
      this.previousTheme = null;
      this.nextTheme = null;
      onComplete?.();
    }, 200);
  }
}

// MARK: - Hook for React Components

/**
 * Hook for using theme transitions in React components
 */
export const useThemeTransition = () => {
  const [transitionManager] = useState(() => new ThemeTransitionManager());
  const [transitionState, setTransitionState] = useState<ThemeTransitionState>(
    transitionManager.getTransitionState()
  );

  useEffect(() => {
    // Update state when transition state changes
    const interval = setInterval(() => {
      setTransitionState(transitionManager.getTransitionState());
    }, 16); // ~60fps

    return () => {
      clearInterval(interval);
      transitionManager.cleanup();
    };
  }, [transitionManager]);

  const transitionToTheme = useCallback(
    (
      currentTheme: Theme,
      newTheme: Theme,
      type: ThemeTransitionType = ThemeTransitionType.FADE,
      curve: ThemeAnimationCurve = ThemeAnimationCurve.EASE_IN_OUT,
      direction: TransitionDirection = TransitionDirection.RIGHT,
      onComplete?: () => void
    ) => {
      transitionManager.transitionToTheme(
        currentTheme,
        newTheme,
        type,
        curve,
        direction,
        onComplete
      );
    },
    [transitionManager]
  );

  return {
    transitionState,
    transitionToTheme,
    getTransitionProgress: () => transitionManager.getTransitionProgress(),
    isTransitioning: () => transitionManager.isTransitioning(),
    interpolateColor: (startColor: string, endColor: string, progress: number) =>
      transitionManager.interpolateColor(startColor, endColor, progress),
  };
};

// MARK: - Animation Helpers

/**
 * Create animated style for fade transition
 */
export const createFadeStyle = (progress: Animated.Value) => ({
  opacity: progress,
});

/**
 * Create animated style for slide transition
 */
export const createSlideStyle = (
  progress: Animated.Value,
  direction: TransitionDirection
) => {
  const distance = 100;
  
  const translateX = progress.interpolate({
    inputRange: [0, 1],
    outputRange: direction === TransitionDirection.LEFT 
      ? [-distance, 0] 
      : direction === TransitionDirection.RIGHT 
      ? [distance, 0] 
      : [0, 0],
  });

  const translateY = progress.interpolate({
    inputRange: [0, 1],
    outputRange: direction === TransitionDirection.UP 
      ? [-distance, 0] 
      : direction === TransitionDirection.DOWN 
      ? [distance, 0] 
      : [0, 0],
  });

  return {
    transform: [{ translateX }, { translateY }],
  };
};

/**
 * Create animated style for scale transition
 */
export const createScaleStyle = (progress: Animated.Value) => {
  const scale = progress.interpolate({
    inputRange: [0, 1],
    outputRange: [0.8, 1],
  });

  return {
    transform: [{ scale }],
    opacity: progress,
  };
};

/**
 * Create animated style for morph transition
 */
export const createMorphStyle = (progress: Animated.Value) => {
  const scale = progress.interpolate({
    inputRange: [0, 1],
    outputRange: [0.95, 1],
  });

  return {
    transform: [{ scale }],
  };
};

/**
 * Create animated style for crossfade transition
 */
export const createCrossfadeStyle = (progress: Animated.Value) => ({
  opacity: progress,
});

/**
 * Create animated style for dissolve transition
 */
export const createDissolveStyle = (progress: Animated.Value) => ({
  opacity: progress,
});

/**
 * Get transition style based on type
 */
export const getTransitionStyle = (
  type: ThemeTransitionType,
  progress: Animated.Value,
  direction?: TransitionDirection
) => {
  switch (type) {
    case ThemeTransitionType.FADE:
      return createFadeStyle(progress);
    case ThemeTransitionType.SLIDE:
      return createSlideStyle(progress, direction || TransitionDirection.RIGHT);
    case ThemeTransitionType.SCALE:
      return createScaleStyle(progress);
    case ThemeTransitionType.MORPH:
      return createMorphStyle(progress);
    case ThemeTransitionType.CROSSFADE:
      return createCrossfadeStyle(progress);
    case ThemeTransitionType.DISSOLVE:
      return createDissolveStyle(progress);
    default:
      return createFadeStyle(progress);
  }
}; 

File: apps/swiftui-app/Sources/AetherThemeKit/ThemeTransitionManager.swift
//
//  ThemeTransitionManager.swift
//  Aether SwiftUI App
//
//  Dynamic theme transitions with smooth animations and system integration.
//  This manager provides polished, delightful theme switching experiences
//  that respect user preferences and accessibility settings.
//

import SwiftUI
import Combine

// MARK: - Transition Types

/// Types of theme transitions available
enum ThemeTransitionType: String, CaseIterable {
    case fade = "Fade"
    case slide = "Slide"
    case scale = "Scale"
    case morph = "Morph"
    case crossfade = "Crossfade"
    case dissolve = "Dissolve"
    
    var description: String {
        switch self {
        case .fade:
            return "Smooth fade between themes"
        case .slide:
            return "Slide transition with direction"
        case .scale:
            return "Scale and fade combination"
        case .morph:
            return "Morphing color transitions"
        case .crossfade:
            return "Crossfade with blur effect"
        case .dissolve:
            return "Dissolve with particle effect"
        }
    }
}

// MARK: - Animation Curves

/// Predefined animation curves for theme transitions
enum ThemeAnimationCurve: String, CaseIterable {
    case easeInOut = "Ease In Out"
    case easeIn = "Ease In"
    case easeOut = "Ease Out"
    case spring = "Spring"
    case bouncy = "Bouncy"
    case smooth = "Smooth"
    
    var animation: Animation {
        switch self {
        case .easeInOut:
            return .easeInOut(duration: 0.6)
        case .easeIn:
            return .easeIn(duration: 0.5)
        case .easeOut:
            return .easeOut(duration: 0.5)
        case .spring:
            return .spring(response: 0.6, dampingFraction: 0.8, blendDuration: 0.3)
        case .bouncy:
            return .spring(response: 0.4, dampingFraction: 0.6, blendDuration: 0.2)
        case .smooth:
            return .interpolatingSpring(stiffness: 100, damping: 10)
        }
    }
}

// MARK: - Transition State

/// State tracking for theme transitions
class ThemeTransitionState: ObservableObject {
    @Published var isTransitioning = false
    @Published var transitionProgress: Double = 0.0
    @Published var currentTransitionType: ThemeTransitionType = .fade
    @Published var currentAnimationCurve: ThemeAnimationCurve = .easeInOut
    @Published var transitionDirection: Edge = .trailing
    
    // Transition timing
    let transitionDuration: Double = 0.6
    let morphingDuration: Double = 0.8
    
    // Accessibility support
    var shouldReduceMotion: Bool {
        UIAccessibility.isReduceMotionEnabled
    }
    
    var shouldReduceTransparency: Bool {
        UIAccessibility.isReduceTransparencyEnabled
    }
}

// MARK: - Theme Transition Manager

/// Manager for handling smooth theme transitions with animations
class ThemeTransitionManager: ObservableObject {
    @Published var transitionState = ThemeTransitionState()
    @Published var previousTheme: Theme?
    @Published var nextTheme: Theme?
    
    private var cancellables = Set<AnyCancellable>()
    private let themeManager: ThemeManager
    
    init(themeManager: ThemeManager) {
        self.themeManager = themeManager
        setupObservers()
    }
    
    // MARK: - Setup
    
    private func setupObservers() {
        // Observe theme changes
        themeManager.$currentTheme
            .sink { [weak self] newTheme in
                self?.handleThemeChange(to: newTheme)
            }
            .store(in: &cancellables)
        
        // Observe accessibility changes
        NotificationCenter.default.publisher(for: UIAccessibility.reduceMotionStatusDidChangeNotification)
            .sink { [weak self] _ in
                self?.updateAccessibilitySettings()
            }
            .store(in: &cancellables)
    }
    
    // MARK: - Theme Transition Methods
    
    /// Perform a smooth theme transition
    /// - Parameters:
    ///   - to: Target theme
    ///   - type: Transition type
    ///   - curve: Animation curve
    ///   - direction: Transition direction (for slide transitions)
    func transitionToTheme(
        _ theme: Theme,
        type: ThemeTransitionType = .fade,
        curve: ThemeAnimationCurve = .easeInOut,
        direction: Edge = .trailing
    ) {
        guard !transitionState.isTransitioning else { return }
        
        // Store transition parameters
        previousTheme = themeManager.currentTheme
        nextTheme = theme
        transitionState.currentTransitionType = type
        transitionState.currentAnimationCurve = curve
        transitionState.transitionDirection = direction
        
        // Check accessibility settings
        if transitionState.shouldReduceMotion {
            // Instant transition for reduced motion
            performInstantTransition(to: theme)
        } else {
            // Animated transition
            performAnimatedTransition(to: theme, type: type, curve: curve)
        }
    }
    
    /// Perform an instant transition (for accessibility)
    private func performInstantTransition(to theme: Theme) {
        withAnimation(.none) {
            themeManager.currentTheme = theme
            transitionState.isTransitioning = false
            transitionState.transitionProgress = 1.0
        }
        
        // Reset after a brief delay
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
            self.transitionState.transitionProgress = 0.0
            self.previousTheme = nil
            self.nextTheme = nil
        }
    }
    
    /// Perform an animated transition
    private func performAnimatedTransition(
        to theme: Theme,
        type: ThemeTransitionType,
        curve: ThemeAnimationCurve
    ) {
        transitionState.isTransitioning = true
        transitionState.transitionProgress = 0.0
        
        // Start transition animation
        withAnimation(curve.animation) {
            transitionState.transitionProgress = 1.0
        }
        
        // Apply theme change at midpoint for smooth transition
        DispatchQueue.main.asyncAfter(deadline: .now() + curve.animation.duration * 0.5) {
            self.themeManager.currentTheme = theme
        }
        
        // Complete transition
        DispatchQueue.main.asyncAfter(deadline: .now() + curve.animation.duration) {
            self.completeTransition()
        }
    }
    
    /// Complete the transition and cleanup
    private func completeTransition() {
        withAnimation(.easeOut(duration: 0.2)) {
            transitionState.isTransitioning = false
        }
        
        // Reset state
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.2) {
            self.transitionState.transitionProgress = 0.0
            self.previousTheme = nil
            self.nextTheme = nil
        }
    }
    
    /// Handle theme change from theme manager
    private func handleThemeChange(to theme: Theme) {
        // This is called when theme manager changes the theme
        // We can add additional transition logic here if needed
    }
    
    /// Update accessibility settings
    private func updateAccessibilitySettings() {
        // Adjust transition behavior based on accessibility settings
        if transitionState.shouldReduceMotion {
            transitionState.currentAnimationCurve = .easeInOut
        }
    }
    
    // MARK: - Transition Effects
    
    /// Get transition effect for a specific transition type
    func transitionEffect(for type: ThemeTransitionType) -> some ViewModifier {
        switch type {
        case .fade:
            return FadeTransitionModifier(progress: transitionState.transitionProgress)
        case .slide:
            return SlideTransitionModifier(
                progress: transitionState.transitionProgress,
                direction: transitionState.transitionDirection
            )
        case .scale:
            return ScaleTransitionModifier(progress: transitionState.transitionProgress)
        case .morph:
            return MorphTransitionModifier(progress: transitionState.transitionProgress)
        case .crossfade:
            return CrossfadeTransitionModifier(progress: transitionState.transitionProgress)
        case .dissolve:
            return DissolveTransitionModifier(progress: transitionState.transitionProgress)
        }
    }
    
    /// Get interpolated color between two themes
    func interpolatedColor(
        from startColor: Color,
        to endColor: Color,
        progress: Double
    ) -> Color {
        let interpolated = startColor.interpolated(to: endColor, amount: progress)
        return interpolated
    }
}

// MARK: - Transition Modifiers

/// Fade transition modifier
struct FadeTransitionModifier: ViewModifier {
    let progress: Double
    
    func body(content: Content) -> some View {
        content
            .opacity(progress)
            .animation(.easeInOut(duration: 0.6), value: progress)
    }
}

/// Slide transition modifier
struct SlideTransitionModifier: ViewModifier {
    let progress: Double
    let direction: Edge
    
    func body(content: Content) -> some View {
        content
            .offset(offset)
            .animation(.easeInOut(duration: 0.6), value: progress)
    }
    
    private var offset: CGSize {
        let distance: CGFloat = 100
        switch direction {
        case .leading:
            return CGSize(width: -distance * (1 - progress), height: 0)
        case .trailing:
            return CGSize(width: distance * (1 - progress), height: 0)
        case .top:
            return CGSize(width: 0, height: -distance * (1 - progress))
        case .bottom:
            return CGSize(width: 0, height: distance * (1 - progress))
        }
    }
}

/// Scale transition modifier
struct ScaleTransitionModifier: ViewModifier {
    let progress: Double
    
    func body(content: Content) -> some View {
        content
            .scaleEffect(scale)
            .opacity(progress)
            .animation(.spring(response: 0.6, dampingFraction: 0.8), value: progress)
    }
    
    private var scale: CGFloat {
        return 0.8 + (0.2 * progress)
    }
}

/// Morph transition modifier
struct MorphTransitionModifier: ViewModifier {
    let progress: Double
    
    func body(content: Content) -> some View {
        content
            .blur(radius: blurRadius)
            .scaleEffect(scale)
            .animation(.easeInOut(duration: 0.8), value: progress)
    }
    
    private var blurRadius: CGFloat {
        return 2.0 * (1 - progress)
    }
    
    private var scale: CGFloat {
        return 0.95 + (0.05 * progress)
    }
}

/// Crossfade transition modifier
struct CrossfadeTransitionModifier: ViewModifier {
    let progress: Double
    
    func body(content: Content) -> some View {
        content
            .blur(radius: blurRadius)
            .opacity(progress)
            .animation(.easeInOut(duration: 0.7), value: progress)
    }
    
    private var blurRadius: CGFloat {
        return 1.0 * (1 - progress)
    }
}

/// Dissolve transition modifier
struct DissolveTransitionModifier: ViewModifier {
    let progress: Double
    
    func body(content: Content) -> some View {
        content
            .opacity(progress)
            .blur(radius: dissolveBlur)
            .animation(.easeInOut(duration: 0.6), value: progress)
    }
    
    private var dissolveBlur: CGFloat {
        return 0.5 * (1 - progress)
    }
}

// MARK: - Color Extensions

extension Color {
    /// Interpolate between two colors
    func interpolated(to other: Color, amount: Double) -> Color {
        // This is a simplified interpolation
        // In a real implementation, you'd want to interpolate RGB/HSB values
        return amount > 0.5 ? other : self
    }
}

// MARK: - Animation Extensions

extension Animation {
    var duration: Double {
        // Extract duration from animation
        // This is a simplified implementation
        return 0.6
    }
}

// MARK: - View Extensions

extension View {
    /// Apply theme transition effect
    func themeTransition(
        _ type: ThemeTransitionType,
        progress: Double
    ) -> some View {
        switch type {
        case .fade:
            return self.opacity(progress)
        case .slide:
            return self.offset(x: 100 * (1 - progress))
        case .scale:
            return self.scaleEffect(0.8 + (0.2 * progress))
        case .morph:
            return self.blur(radius: 2.0 * (1 - progress))
        case .crossfade:
            return self.blur(radius: 1.0 * (1 - progress))
        case .dissolve:
            return self.opacity(progress).blur(radius: 0.5 * (1 - progress))
        }
    }
} 